File: tests/fxp_sim/loader.py
# -*- coding: utf-8 -*-
"""Carregador do IR no runtime do protótipo — mock do front-end da Etapa 1.

Na Etapa 2 (PLAN §2.1) o lexer/parser produzirá o IR direto do texto `.vl`;
aqui os testes carregam o IR construído por `ir.py`. Toda a semântica de
validação estrutural fica no `contract.py` — carregar um programa inválido
é bug do teste, não do loader.
"""

from __future__ import annotations

from . import ir
from .blueprint import load as _load_prototype


def _vbl():
    return _load_prototype()


def load(engine, program: dict):
    """Carrega formas, reviews e o bloco main no engine.

    Retorna o `MainInterpreter` (ou None se o programa não tiver main).
    """
    for declaration in program["declarations"]:
        if declaration["type"] == "form":
            _load_form(engine, declaration)
        elif declaration["type"] == "review":
            _load_review(engine, declaration)
        else:
            raise ValueError(f"declaração desconhecida: {declaration['type']}")

    block = program.get("main")
    if block is None:
        return None
    return _load_main(engine, block)


# ----------------------------------------------------------------------
def _load_form(engine, decl: dict):
    attributes = decl["attributes"]
    value = attributes["value"]
    horizon = ir.duration(attributes["horizon"])
    source_path = attributes.get("source_path")
    now = engine.sim_time
    name = decl["name"]

    if decl["conjugation"] == "event":
        form_obj = _vbl().EventForm(name, value, horizon, source_path, now)
    elif decl["conjugation"] == "equilibrium":
        form_obj = _vbl().EquilibriumForm(
            name, value, horizon, source_path, attributes.get("cost_bytes"), now
        )
    elif decl["conjugation"] == "nonequilibrium":
        if "maintenance_deadline" not in attributes:
            raise ValueError(
                f"forma '{name}': nonequilibrium exige maintenance_deadline "
                f"(validação estrutural fica no contract.py)"
            )
        form_obj = _vbl().NonequilibriumForm(
            name, value, horizon, source_path,
            ir.duration(attributes["maintenance_deadline"]),
            attributes.get("exchange_mode", "cooperation"), now,
        )
    else:
        raise ValueError(f"conjugação desconhecida: {decl['conjugation']}")

    # opcionais comuns (sobrepõem o padrão da conjugação)
    if "currency" in attributes:
        form_obj.currency = attributes["currency"]
    if "classification" in attributes:
        form_obj.classification_currency = attributes["classification"]

    engine.register_form(form_obj)
    return form_obj


def _load_review(engine, decl: dict):
    form_obj = engine.forms.get(decl["form"])
    if form_obj is None:
        raise ValueError(f"review para forma inexistente: {decl['form']}")
    for rule in decl["rules"]:
        threshold = rule["threshold"]  # unidade vira número puro (FORMAL §3)
        form_obj.add_review_condition(
            rule["sensor"], rule["op"], threshold,
            [_translate_action(a) for a in rule["actions"]],
        )
    return form_obj


def _translate_action(action: dict) -> dict:
    """IR -> estrutura esperada pelo protótipo (actor/value)."""
    return dict(action)


def _load_main(engine, block: dict):
    """Traduz statements do IR para a estrutura esperada pelo MainInterpreter."""
    interpreter = _vbl().MainInterpreter(engine)
    periodic: list[tuple[float, list[dict]]] = []

    def translate(stmt: dict) -> dict:
        kind = stmt["statement"]
        if kind == "keep":
            return {"statement": "keep", "form": stmt["form"]}
        if kind == "act":
            return {"statement": "act", "actor": stmt["actor"], "value": stmt["value"]}
        if kind == "every":
            periodic.append(
                (ir.duration(stmt["period"]), [translate(s) for s in stmt["statements"]])
            )
            return {}
        raise ValueError(f"statement desconhecido: {kind}")

    direct = []
    for stmt in block["statements"]:
        translated = translate(stmt)
        if translated:
            direct.append(translated)
    if direct:
        # statements de topo do main rodam como um bloco `every` de período 1 tick
        periodic.insert(0, (engine.tick_seconds, direct))
    for period, statements in periodic:
        interpreter.add_every(period, statements)
    return interpreter

File: core/crates/vbl-runtime/src/loader.rs
//! Carregador: AST → runtime (formas, reviews e bloco `main`).
//!
//! Equivalente Rust do `loader.py` + `contract.py` da Etapa 1: carrega o
//! programa no engine e valida referências contra o REGISTRO do FXP
//! (sensor/ator registrados, unidade compatível com a grandeza — FORMAL §3/§6).
//! As cláusulas estruturais já são erros de compilação do parser (`vbl-lang`).
//!
//! `carregar` NÃO exige registro válido (o runtime lida com falhas de I/O
//! por §4.7 — sensor ausente nunca dispara regra); `validar` devolve os
//! diagnósticos de registro para o `check` do CLI.

use crate::notebook::Caderno;
use crate::engine::Engine;
use crate::form::{ActionRt, Form, Manutencao, RuleRt};
use crate::fxp::{Fxp, Value};
use crate::main_interp::{MainInterpreter, StmtRt};
use vbl_lang::{Action, Conjugation, Declaration, ExprKind, Program, Statement};

/// Diagnóstico de carga (referências contra o registro do FXP).
#[derive(Debug, Clone)]
pub struct LoadDiag {
    pub code: String,
    pub message: String,
}

impl std::fmt::Display for LoadDiag {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "[{}] {}", self.code, self.message)
    }
}

/// Carrega o programa no engine (formas na ordem de declaração, regras na
/// ordem declarada, bloco `main` interpretado). Devolve o interpretador do
/// `main` (vazio se o programa não tiver `main`).
pub fn carregar<F: Fxp, C: Caderno>(
    engine: &mut Engine<F, C>,
    programa: &Program,
) -> MainInterpreter {
    for decl in &programa.decls {
        match decl {
            Declaration::Form(f) => engine.registrar_forma(forma_runtime(f, engine.sim_time)),
            Declaration::Review(r) => {
                for regra in &r.rules {
                    let acoes: Vec<ActionRt> = regra.actions.iter().map(acao_runtime).collect();
                    let rt = RuleRt {
                        sensor: regra.sensor.nome.clone(),
                        op: regra.op,
                        threshold: regra.threshold.valor, // unidade vira número puro (FORMAL §3)
                        actions: acoes,
                    };
                    engine.forma_mut(&r.form).unwrap_or_else(|| {
                        panic!(
                            "review sem forma: '{}' — cláusula de compilação (review_orfa) não verificada",
                            r.form
                        )
                    })
                    .rules
                    .push(rt);
                }
            }
        }
    }
    let mut interp = MainInterpreter::default();
    if let Some(main) = &programa.main {
        let mut diretos: Vec<StmtRt> = Vec::new();
        for st in &main.statements {
            match statement_runtime(st) {
                StmtKind::Direto(s) => diretos.push(s),
                StmtKind::Every(periodo_s, body) => interp.add_every(periodo_s, body),
            }
        }
        if !diretos.is_empty() {
            // statements de topo rodam como bloco `every` de 1 tick (Etapa 1)
            interp.add_every(engine.tick_seconds(), diretos);
        }
    }
    interp
}

enum StmtKind {
    Direto(StmtRt),
    Every(f64, Vec<StmtRt>),
}

fn statement_runtime(st: &Statement) -> StmtKind {
    match st {
        Statement::Keep(forma) => StmtKind::Direto(StmtRt::Keep(forma.clone())),
        Statement::Act { actor, value } => StmtKind::Direto(StmtRt::Act {
            ator: actor.clone(),
            valor: expr_value(value),
        }),
        Statement::Every { period, body } => StmtKind::Every(
            period.segundos(),
            body.iter().map(statement_runtime_direto).collect(),
        ),
    }
}

fn statement_runtime_direto(st: &Statement) -> StmtRt {
    match statement_runtime(st) {
        StmtKind::Direto(s) => s,
        StmtKind::Every(..) => unreachable!("every aninhado tratado como direto"),
    }
}

fn expr_value(expr: &vbl_lang::Expression) -> Value {
    match &expr.kind {
        ExprKind::Str(s) => Value::Str(s.clone()),
        ExprKind::Num(n) => Value::Num(*n),
        ExprKind::Ident(s) => Value::Ident(s.clone()),
    }
}

fn acao_runtime(a: &Action) -> ActionRt {
    match a {
        Action::Dissolve => ActionRt::Dissolve,
        Action::Subvert => ActionRt::Subvert,
        Action::ReclassifyAsEquilibrium => ActionRt::ReclassifyEquilibrium,
        Action::ReclassifyAsNonequilibrium => ActionRt::ReclassifyNonequilibrium,
        Action::NotifyShutdown => ActionRt::NotifyShutdown,
        Action::Act { actor, value, .. } => {
            ActionRt::Act { ator: actor.clone(), valor: expr_value(value) }
        }
    }
}

/// Converte a declaração de forma para a forma runtime (criação em `agora`).
pub fn forma_runtime(f: &vbl_lang::FormDecl, agora: f64) -> Form {
    let horizon = f.horizon.segundos();
    let mut form = Form {
        name: f.name.clone(),
        value: expr_value(&f.value),
        horizon_s: horizon,
        creation_time: agora,
        conjugation: f.conjugation,
        currency: f
            .attrs
            .currency
            .clone()
            .unwrap_or_else(|| f.conjugation.currency_padrao().into()),
        source_path: f.attrs.source_path.clone(),
        classification: f.attrs.classification.clone(),
        declared_maintenance_deadline: None,
        manutencao: None,
        exchange_mode: None,
        cost_bytes: None,
        rules: Vec::new(),
        dissolvida: false,
        horizon_versao: 0,
        manutencao_versao: 0,
    };
    match f.conjugation {
        Conjugation::Nonequilibrium => {
            let deadline = f
                .attrs
                .maintenance_deadline
                .as_ref()
                .map(|d| d.segundos())
                .unwrap_or_else(|| {
                    // cláusula de compilação (maintenance_deadline_ausente);
                    // defesa: programa inválido não chega aqui pelo CLI
                    panic!("nonequilibrium '{}' sem maintenance_deadline", f.name)
                });
            form.declared_maintenance_deadline = Some(deadline);
            form.manutencao = Some(Manutencao { deadline_s: deadline, ultima: agora });
            form.exchange_mode =
                Some(f.attrs.exchange_mode.clone().unwrap_or_else(|| "cooperation".into()));
        }
        Conjugation::Equilibrium => {
            form.cost_bytes = f.attrs.cost_bytes.map(|b| b.max(0) as u64);
        }
        Conjugation::Event => {}
    }
    form
}

// ------------------------------------------------------------------
// Validação contra o registro do FXP (contract.py da Etapa 1)
// ------------------------------------------------------------------
/// Valida referências do programa contra o registro do FXP. Devolve apenas
/// diagnósticos de REGISTRO (os estruturais são do parser).
pub fn validar(registry: &crate::fxp::Registry, programa: &Program) -> Vec<LoadDiag> {
    let mut diags = Vec::new();
    for f in programa.forms() {
        if let Some(sp) = &f.attrs.source_path {
            if !registry.sensores.contains_key(sp) {
                diags.push(LoadDiag {
                    code: "sensor_nao_registrado".into(),
                    message: format!("source_path '{sp}' fora do registro do FXP"),
                });
            }
        }
    }
    for r in programa.reviews() {
        for (i, regra) in r.rules.iter().enumerate() {
            let Some(info) = registry.sensores.get(&regra.sensor.nome) else {
                diags.push(LoadDiag {
                    code: "sensor_nao_registrado".into(),
                    message: format!(
                        "review {} regra#{}: sensor '{}' fora do registro",
                        r.form, i, regra.sensor.nome
                    ),
                });
                continue;
            };
            if let Some(unit) = regra.threshold.unit {
                // unidade compatível com a grandeza do sensor (FORMAL §3)
                if !unit.grandeza().is_empty()
                    && !info.grandeza.is_empty()
                    && unit.grandeza() != info.grandeza
                {
                    diags.push(LoadDiag {
                        code: "unidade_incompativel".into(),
                        message: format!(
                            "review {} regra#{}: unidade '{}' incompatível com a grandeza '{}' do sensor '{}' (esperado por grandeza: '{}')",
                            r.form,
                            i,
                            unit.simbolo(),
                            info.grandeza,
                            regra.sensor.nome,
                            info.unidade
                        ),
                    });
                }
            } else {
                diags.push(LoadDiag {
                    code: "unidade_ausente".into(),
                    message: format!(
                        "review {} regra#{}: sensor '{}' tem grandeza '{}' — threshold exige unidade ({}), FORMAL §3/§6",
                        r.form, i, regra.sensor.nome, info.grandeza, info.unidade
                    ),
                });
            }
            for acao in &regra.actions {
                if let Action::Act { actor, .. } = acao {
                    if !registry.atores.contains_key(actor) {
                        diags.push(LoadDiag {
                            code: "ator_nao_registrado".into(),
                            message: format!(
                                "review {} regra#{}: ator '{actor}' fora do registro FXP",
                                r.form, i
                            ),
                        });
                    }
                }
            }
        }
    }
    if let Some(main) = &programa.main {
        fn passe(stmts: &[Statement], registry: &crate::fxp::Registry, diags: &mut Vec<LoadDiag>) {
            for st in stmts {
                match st {
                    Statement::Act { actor, .. } => {
                        if !registry.atores.contains_key(actor) {
                            diags.push(LoadDiag {
                                code: "ator_nao_registrado".into(),
                                message: format!("main: ator '{actor}' fora do registro FXP"),
                            });
                        }
                    }
                    Statement::Every { body, .. } => passe(body, registry, diags),
                    Statement::Keep(_) => {}
                }
            }
        }
        passe(&main.statements, registry, &mut diags);
    }
    diags
}
